rendezvous/README.md

```diff
@@ -0,0 +1,106 @@
+# Rendezvous Service
+
+Scope:
+- real-time applications that require rendezvous
+- replace ws-star-rendezvous with a rendezvous service daemon and a fleet
+  of p2p-circuit relays.
+
+## Rendezvous Protocol
+
+The rendezvous protocol provides facilities for real-time peer discovery within
+application specific namespaces. Peers connect to the rendezvous service and register
+their presence in one or more namespaces. Registrations persist until the peer
+disconnects or explicitly unregisters.
+
+Peers can enter rendezvous and dynamically receive announcements about peer
+registrations and unregistrations within their namespaces of interest.
+For purposes of oneof discovery (eg bootstrap), peers can also ask the service
+for a list of peers within a namespace.
+
+### Interaction
+
+Client peer `A` connects to the rendezvous service `R` and registers for namespace
+`my-app` with a `REGISTER` message. It subsequently enters rendezvous with
+a `RENDEZVOUS` message and waits for `REGISTER`/`UNREGISTER` announcements from
+the service.
+
+```
+A -> R: REGISTER{my-app, {QmA, AddressA}}
+A -> R: RENDEZVOUS{my-app}
+```
+
+Client peer `B` connects, registers and enters rendezvous.
+The rendezvous service immediately notifies `B` about the current namespace registrations
+and emits a register notification to `A`:
+
+```
+B -> R: REGISTER{my-app, {QmB, AddressB}}
+B -> R: RENDEZVOUS{my-app}
+
+R -> B: REGISTER{my-app, {QmA, AddressA}}
+R -> A: REGISTER{my-app, {QmB, AddressB}}
+```
+
+A third client `C` connections and registers:
+```
+C -> R: REGISTER{my-app, {QmC, AddressC}}
+C -> R: RENDEZVOUS{my-app}
+
+R -> C: REGISTER{my-app, {QmA, AddressA}}
+        REGISTER{my-app, {QmB, AddressB}}
+R -> A: REGISTER{my-app, {QmC, AddressC}}
+R -> B: REGISTER{my-app, {QmC, AddressC}}
+```
+
+A client can discover peers in the namespace by sending a `DISCOVER` message; the
+service responds with the list of current peer reigstrations.
+```
+D -> R: DISCOVER{my-app}
+R -> D: REGISTER{my-app, {QmA, AddressA}}
+        REGISTER{my-app, {QmB, AddressB}}
+        REGISTER{my-app, {QmC, AddressC}}
+```
+
+### Protobuf
+
+
+```protobuf
+message Message {
+  enum MessageType {
```

Comment:
Wouldn't it make more sense to merge register and unregister requests? Or dropping the message type completly?
The client could simply check the length of the array for all types.

Reply:
I think we want unregister so that we can depart from a namespace without having to disconnect.
The message type is needed to disambiguate the protocol.